Objects/abstract.c

```diff
@@ -2468,31 +2468,71 @@ PyMapping_HasKeyWithError(PyObject *obj, PyObject *key)
 }
 
 int
-PyMapping_HasKeyString(PyObject *o, const char *key)
+PyMapping_HasKeyString(PyObject *obj, const char *key)
 {
-    PyObject *v;
-
-    v = PyMapping_GetItemString(o, key);
-    if (v) {
-        Py_DECREF(v);
-        return 1;
+    PyObject *value;
+    int rc;
+    if (obj == NULL) {
+        // For backward compatibility.
+        // PyMapping_GetOptionalItemString() crashes if obj is NULL.
+        null_error();
+        rc = -1;
     }
-    PyErr_Clear();
-    return 0;
+    else {
+        rc = PyMapping_GetOptionalItemString(obj, key, &value);
+    }
+    if (rc < 0) {
+        PyErr_FormatUnraisable(
+            "Exception ignored in PyMapping_HasKeyString(); consider using "
+            "PyMapping_HasKeyStringWithError(), "
+            "PyMapping_GetOptionalItemString() or PyMapping_GetItemString()");
+        return 0;
+    }
+    // PyMapping_HasKeyString() also clears the error set before it's called
+    // if the key is not found.
+    if (rc == 0 && PyErr_Occurred()) {
+        PyErr_FormatUnraisable(
+            "Ignore exception set before calling in PyMapping_HasKeyString(); "
+            "consider using PyMapping_HasKeyStringWithError(), "
+            "PyMapping_GetOptionalItemString() or PyMapping_GetItemString()");
```

Comment:
These 3 functions **can** be called with an exception set, and they don't override the exception? That sounds surprising. I would prefer suggesting to not call the function with an exception set. What do you think?

Reply:
Got catch. Indeed, `PyMapping_GetOptionalItemString()` can only return 0 if no exception is set, so this condition is always false. Also, the alternative functions also can clear exceptions.

I think that we should classify the C API by classes:

1. Function that can be called when an exception is set, and they do not change it.
2. Function that can be called when an exception is set, and they do not change it unless they fail.
3. Function that can be called when an exception is set, but they can change it even at success.
4. Function that can be called when an exception is set, but the result is ambiguous in some cases (you cannot distinguish some successful results from failure).
5. Function that should never be called when an exception is set.

There may be more classes.

Reply:
In the general case, to write safe code handling a raised exception, I think the safest option is to keep the exception aside using `PyErr_GetRaisedException()`. Maybe today some functions are perfectly fine and never override the currently raised exception. But what if tomorrow their implementation changes, and they may start to clear the currently raised exception?

In Python, in an ``except:`` block, there is no "currently raised exception" in the C level, even if `sys.exc_info()` returns the exception. The difference between `PyThreadState.exc_info` and `PyThreadState.current_exception` is subtle.

Reply:
This is why it should be clearly documented.

Obviously you can call `PyErr_Occurred()`, `PyErr_GetRaisedException()` or `PyErr_WriteUnraisable()` when an exception is set.